src/transforms/key_value_parser.rs

```diff
@@ -0,0 +1,409 @@
+        let field = self
+            .field
+            .as_ref()
+            .unwrap_or(&event::log_schema().message_key());
+
+        let conversions = parse_conversion_map(&self.types)?;
+
+        let trim_key = if let Some(key) = &self.trim_key {
+            Some(key.chars().collect())
+        } else {
+            None
+        };
+
+        let trim_value = if let Some(val) = &self.trim_value {
+            Some(val.chars().collect())
+        } else {
+            None
+        };
+
+        Ok(Box::new(KeyValue {
+            conversions,
+            drop_field: self.drop_field,
+            field: field.clone(),
+            field_split: Option::from(self.field_split.clone()),
+            overwrite_target: self.overwrite_target,
+            separator: self.separator.clone(),
+            target_field: self.target_field.clone(),
+            trim_key,
+            trim_value,
+        }))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "key_value_parser"
+    }
+}
+
+pub struct KeyValue {
+    conversions: HashMap<Atom, Conversion>,
+    drop_field: bool,
+    field: Atom,
+    field_split: Option<String>,
+    overwrite_target: bool,
+    separator: String,
+    target_field: Option<Atom>,
+    trim_key: Option<Vec<char>>,
+    trim_value: Option<Vec<char>>,
+}
+
+impl KeyValue {
+    pub fn parse_pair(&self, pair: &str) -> Option<(Atom, String)> {
+        let pair = pair.trim();
+        let field_split = self
+            .field_split
+            .as_ref()
+            .ok_or_else(|| Some("".to_string()))
+            .unwrap();
+
+        let fields = if field_split.is_empty() {
+            let mut kv_pair = pair.split_whitespace();
+            let key = kv_pair.next()?;
+            let val = kv_pair.next()?;
+            if kv_pair.next().is_some() {
+                error!(
+                    message = "KV parser saw more than one separator",
+                    rate_limit_secs = 30
+                );
+                return None;
+            }
+
+            (key, val)
+        } else {
+            let split_index = pair.find(field_split).unwrap_or(0);
+            let (key, _val) = pair.split_at(split_index);
+            let key = key.trim();
+            let val = pair[split_index + field_split.len()..].trim();
+
+            if key.is_empty() || val.is_empty() {
+                return None;
+            }
+
+            (key, val)
+        };
+
+        let key = if let Some(trim_key) = &self.trim_key {
+            fields.0.trim_matches(trim_key as &[_])
+        } else {
+            fields.0
+        };
+
+        let val = if let Some(trim_value) = &self.trim_value {
+            fields.1.trim_matches(trim_value as &[_])
+        } else {
+            fields.1
+        };
```

Comment:
Ditto match

Reply:
It seems I found an `if let Some` hammer and over used it on my last round of updates, haha. Ditto for the previous Ditto reply; what are some things you look at and take into consideration what making decisions in these situations where there is more than one way to do it in Rust?

Reply:
I tend to use a few relatively simple metrics to guide me:
1. What is the highest-level language construct for this task? (functional operators >= `match` >= `let X = if … else …` >= `let mut X; if _ { X = … } else { X = … }`)
1. What construct most clearly represents what I want to achieve (as opposed to what the compiler will translate it into)?
1. What results in the most compact representation of the semantics?

For `if let` in particular, if it is balanced by an `else` clause, I will almost always use `match` instead as it better represents what is being done (there are a few cases where `rustfmt` will turn the `if` into a single line but the `match` will always be multi-line). Similarly for pattern matching, using `match` across all the values that need matching lets us be sure all the cases are enumerated, and the compiler is free to optimize it into nested `if/else` blocks internally. Further, if the operation can be represented by functional operators (`.map`, `.and_then`, etc), I will tend to prefer those. In this case it would be possible to use them, but it doesn't clarify anything:
```rust
    let val = self
        .trim_value
        .as_ref()
        .map(|trim_value| fields.1.trim_matches(trim_value as &[_]))
        .unwrap_or(fields.1);
```

Reply:
This is awesome, and very helpful, thanks!